packages/@lwc/ssr-runtime/src/clone-and-deep-freeze.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { freeze, entries, isObject, isArray, create, isNull } from '@lwc/shared';
+
+// Deep freeze and clone an object. Designed for cloning/freezing child props when passed from a parent to a child so
+// that they are immutable. This is one of the normal guarantees of both engine-dom and engine-server that we want to
+// emulate in ssr-runtime. The goal here is that a child cannot mutate the props of its parent and thus affect
+// the parent's rendering, which would lead to bidirectional reactivity and mischief.
+export function cloneAndDeepFreeze(obj: any): any {
```

Comment:
Could something like `immer` handle this? (Do we have that in runtime or just compiler?)

Reply:
I think `immer` is just in the compiler.

Long-term, we may delete this function entirely. I'm sure it adds perf overhead, and it's arguably not useful since you will likely get an error on the client side if you rely on children mutating their parents' props. Maybe it makes sense in dev-only, or pure-SSR-(no hydration)-only.